pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -944,77 +944,94 @@ func (c *controller) setMachineTerminationStatus(ctx context.Context, deleteMach
 	return machineutils.ShortRetry, err
 }
 
-// getVMStatus tries to retrive VM status backed by machine
+// getVMStatus tries to retrieve VM status backed by machine
 func (c *controller) getVMStatus(ctx context.Context, getMachineStatusRequest *driver.GetMachineStatusRequest) (machineutils.RetryPeriod, error) {
 	var (
 		retry       machineutils.RetryPeriod
 		description string
 		state       v1alpha1.MachineState
+		err         error
 	)
-
-	statusResp, err := c.driver.GetMachineStatus(ctx, getMachineStatusRequest)
-	if err == nil {
-		// VM Found
-
-		// If `node` label is missing on machine obj, then update this label on Machine object with nodeName from status response
-		nodeName := getMachineStatusRequest.Machine.Labels[v1alpha1.NodeLabelKey]
-		if nodeName == "" {
-			err = c.updateMachineNodeLabel(ctx, getMachineStatusRequest.Machine, statusResp.NodeName)
-			if err != nil {
-				return machineutils.ShortRetry, err
+	// Node label is required for drain of node, therefore we try to update machine object before proceeding to drain.
+	isNodeLabelUpdated := false
+	//check if node name label is already present in machine object
+	nodeName := getMachineStatusRequest.Machine.Labels[v1alpha1.NodeLabelKey]
+	if isValidNodeName(nodeName) {
+		isNodeLabelUpdated = true
+	} else {
+		// Figure out node label either by checking all nodes for label matching machine name or retrieving it using GetMachineStatus
+		// get all nodes and check if any node has the machine name as label
+		var nodes []*v1.Node
+		nodes, err = c.nodeLister.List(labels.Everything())
+		if err == nil {
+			for _, node := range nodes {
+				if node.Labels["node.gardener.cloud/machine-name"] == getMachineStatusRequest.Machine.Name {
+					nodeName := node.Name
+					err = c.updateMachineNodeLabel(ctx, getMachineStatusRequest.Machine, nodeName)
+					if err != nil {
+						return machineutils.ShortRetry, err
+					}
+					isNodeLabelUpdated = true
+					break
+				}
 			}
 		}
-
+		if !isNodeLabelUpdated {
```

Comment:
It would be better to rewrite this function.
An idea:
```go
func (c *controller) updateMachineNodeLabelAndStatus(ctx context.Context, machine *v1alpha1.Machine) {
   // calls getNodeName
   // if there is no error then updates the node label
   // if there is an error either in calling getNodeName or updating the node label then updates the status accordingly

}

func (c *controller) getNodeName(ctx context.Context, machine *v1alpha1.Machine) (string, error) {
	nodeName := machine.Labels[v1alpha1.NodeLabelKey]
	if len(strings.TrimSpace(nodeName)) > 0 {
		return nodeName, nil
	}
	matchingNode, err := c.fetchMatchingNode(ctx, machine.Name)	
	if err == nil && matchingNode != nil {
		return matchingNode.Name, nil
	}
	if err != nil {
		klog.Errorf("Error trying to get node matching machine %s: %v. Will try to get the node name by calling driver.GetMachineStatus instead.", machine.Name, err)
	}
	// call GetMachineStatus to get the node name	
}

```